### PR TITLE
Stat: Add Percent Change Option

### LINKED ISSUE #5959: Feature Request: Single Stat Trends
- *_I'm submitting a [ X ] Feature request *_
  for Grafana 3.1.1 influx/graphite

it would be cool to have in the single stat (or in another similar widget), the ability to summarize the trend of the KPI.

like you want to monitor the number of download per day...if i display the number , say 42, it won't say much .... but if you include an arrow up or maybe +42% it can have really much more value.

The trend can be computed over the half the period (first half for previous number, second half for the current number) or over the same amount of time before the period or smth like this

visually it could be represented like this

capture d ecran 2016-09-03 a 14 52 11
capture d ecran 2016-09-03 a 14 51 56
capture d ecran 2016-09-03 a 14 43 23
capture d ecran 2016-09-03 a 14 53 40

### PR DESCRIPTION
**Pull request by drew08t**
Add an option to display percent change on stat panel.

- [x] General styling work
- [x] Improve styling approach to be consistent with [current work](https://github.com/grafana/grafana/pull/78032)

Before:
image
image

After:
image
image
image

Closes #5959

### CODE REVIEW

**Comment by drew08t** [comment]
Perhaps chart up/down icons would fit better.
image

File: packages/grafana-ui/src/components/BigValue/BigValue.tsx [hunk 1]
```diff
@@ -103,7 +111,22 @@ export class BigValue extends PureComponent<Props> {
         <div className={className} style={panelStyles} title={tooltip}>
           <div style={valueAndTitleContainerStyles}>
             {textValues.title && <div style={titleStyles}>{textValues.title}</div>}
-            <FormattedValueDisplay value={textValues} style={valueStyles} />
+            <div style={valueContainerStyles}>
```
Comment (nmarrs) [inline packages/grafana-ui/src/components/BigValue/BigValue.tsx:114]: Make this so that the new style is only applied when `percentChange` is enabled

**Review by nmarrs — COMMENTED** [review]
We should add a section in gdev dashboard for this case

**Comment by lukasztyrala** [comment]
This is an excellent first step. However, I have a few concerns and questions:
1. Broader context:
- Adding the trend indicator and #78032 introduces a lot of permutations to the possible layouts.
- The interplay of `Orientation`, `Text mode`, `Graph mode`, `Text alignment`, `Unit`, `Text size` is already not super intuitive, and requires complex logic for displaying things smartly.
2. This change:
- Placing the trend indicator after the value with units (especially long ones) can break the layout fast.
- Displaying arrow indicator above percentage change, when sizes are altered can result it such a grouping that it will be hard for people to connect the arrow to percentages correctly.
- IMHO the arrow is fine for now.
- What additional option will we introduce with the trend indicator?
- What would be the trend indicator for no change (I assume `-` but it would be good to test it out)?
- How scaling will affect the placement and size?
- Maybe we could try to place the arrow + percentage below the value?

Screenshot 2023-11-22 at 15 05 25

I'd be happy to explore the permutations and possible layouts, but I will need a day or two for that :-)

**Comment by drew08t** [comment]
> This is an excellent first step. However, I have a few concerns and questions:
> 
> 1. Broader context:
> 
> * Adding the trend indicator and [Stat: Minor wide layout option cleanup #78032](https://github.com/grafana/grafana/pull/78032) introduces a lot of permutations to the possible layouts.
> * The interplay of `Orientation`, `Text mode`, `Graph mode`, `Text alignment`, `Unit`, `Text size` is already not super intuitive, and requires complex logic for displaying things smartly.
> 
> 2. This change:
> 
> * Placing the trend indicator after the value with units (especially long ones) can break the layout fast.
> * Displaying arrow indicator above percentage change, when sizes are altered can result it such a grouping that it will be hard for people to connect the arrow to percentages correctly.
> * IMHO the arrow is fine for now.
> * What additional option will we introduce with the trend indicator?
> * What would be the trend indicator for no change (I assume `-` but it would be good to test it out)?
> * How scaling will affect the placement and size?
> * Maybe we could try to place the arrow + percentage below the value?
> 
> Screenshot 2023-11-22 at 15 05 25
> I'd be happy to explore the permutations and possible layouts, but I will need a day or two for that :-)

I think horizontally aligned arrow next to the value (as you illustrated in gray rectangles) would look great beneath the primary value or off to the side for wide layout.

File: packages/grafana-ui/src/components/BigValue/BigValue.tsx [hunk 2]
```diff
@@ -87,23 +89,42 @@ export class BigValue extends PureComponent<Props> {
...
+      parentOrientation === VizOrientation.Horizontal ? layout.maxTextHeight / 3 : layout.maxTextHeight / 8;
 
     if (!onClick) {
       return (
         <div className={className} style={panelStyles} title={tooltip}>
           <div style={valueAndTitleContainerStyles}>
             {textValues.title && <div style={titleStyles}>{textValues.title}</div>}
             <FormattedValueDisplay value={textValues} style={valueStyles} />
+            {!percentChangeNaN && textMode !== BigValueTextMode.Name && textMode !== BigValueTextMode.None && (
+              <div style={percentChangeStyles}>
```
Comment (nmarrs) [inline packages/grafana-ui/src/components/BigValue/BigValue.tsx:119]: I think we need to not render this also when `percentChange` is undefined as it seems to be impacting the layout / design even when the percent changed option is set to false - i.e. the value is being pushed over to the left whether or not percent is being displayed


https://github.com/grafana/grafana/assets/22381771/a0dc93d0-d4d3-483f-8d30-848751e8c48f

**Review by nmarrs — COMMENTED** [review]
overall looking good! just a few nits / follow-ups :)

File: packages/grafana-ui/src/components/BigValue/BigValue.tsx [hunk 3]
```diff
@@ -87,23 +89,42 @@ export class BigValue extends PureComponent<Props> {
...
     const valueStyles = layout.getValueStyles();
     const titleStyles = layout.getTitleStyles();
     const textValues = layout.textValues;
+    const percentChange = this.props.value.percentChange;
+    const percentChangeNaN = Number.isNaN(percentChange);
+    const percentChangeString =
+      percentChange?.toLocaleString(undefined, { style: 'percent', maximumSignificantDigits: 3 }) ?? '';
+    const percentChangeIcon =
+      percentChange && (percentChange > 0 ? 'arrow-up' : percentChange < 0 ? 'arrow-down' : undefined);
+    const percentChangeStyles = layout.getValueStyles(true);
```
[the remark this answers is not in the document]
Comment (nmarrs) [reply packages/grafana-ui/src/components/BigValue/BigValue.tsx:105]: breaking this out into separate component would also make it easier to extend in the future if we want to provide more options for % sizing / placement / color etc

File: packages/grafana-ui/src/components/BigValue/BigValue.tsx [hunk 4]
```diff
@@ -87,23 +89,42 @@ export class BigValue extends PureComponent<Props> {
...
 
     // When there is an outer data link this tooltip will override the outer native tooltip
     const tooltip = hasLinks ? undefined : textValues.tooltip;
+    const iconDim =
+      parentOrientation === VizOrientation.Horizontal ? layout.maxTextHeight / 3 : layout.maxTextHeight / 8;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/BigValue/BigValue.tsx:110]: I don't see how VizRepeater orientation matters here (also removing this property in https://github.com/grafana/grafana/pull/79389) 

and this check also makes does work for the "Auto" orientation (where VizRepeater will automatically decide orientation)

File: packages/grafana-ui/src/components/BigValue/BigValueLayout.tsx [hunk 5]
```diff
@@ -77,9 +77,16 @@ export abstract class BigValueLayout {
     return styles;
   }
 
-  getValueStyles(): CSSProperties {
+  getValueStyles(percentChange?: boolean): CSSProperties {
+    const valueFontSize = this.valueFontSize;
+    const fontSize = percentChange
+      ? this.props.parentOrientation === VizOrientation.Horizontal
```
Comment (torkelo) [inline packages/grafana-ui/src/components/BigValue/BigValueLayout.tsx:83]: Can you describe why the VizRepeater orientation matters here? Struggle seeing when we would care, only the orientation (layout) of the BigValue itself should matter?

File: public/app/plugins/panel/stat/module.tsx [hunk 6]
```diff
@@ -87,6 +87,13 @@ export const plugin = new PanelPlugin<Options>(StatPanel)
             { value: BigValueJustifyMode.Center, label: 'Center' },
           ],
         },
+      })
+      .addBooleanSwitch({
+        path: 'showPercentChange',
+        name: 'Show percent change',
+        defaultValue: defaultOptions.showPercentChange,
+        category: mainCategory,
+        showIf: (config) => config.textMode !== BigValueTextMode.Name && config.textMode !== BigValueTextMode.None,
```
Comment (torkelo) [inline public/app/plugins/panel/stat/module.tsx:96]: I could see text mode name and percent change being useful

**Review by torkelo — COMMENTED** [review]
Some layout issues. esp in value only mode:
image

Also unsure why BigValue cares about the VizRepeater layout being hard set to Horizontal in a few places.  removing the parentOrientation in https://github.com/grafana/grafana/pull/79389 

There is also a problem showing this option when we are showing row data (Value options  > All values).  In this mode we need to pick a field to use for the percent change value (but could be done in a follow up PR and percent change option hidden when Show is set to all Values in the mean time). 
 
Screenshot from 2023-12-12 13-07-33

The other big problem here is color by, almost all other visualization tools that show stats with percent change color the percent change value red if negative and green if positive. Here we always use the same color as the value which looks a bit off. 

But tricky, we could hard code coloring of percent change, a bit tricky to allow customization unless we make the change value be presented as a standalone field.,

**Comment by torkelo** [comment]
Oh forgot to say that I am excited we are working on this! This has felt like a major oversight / missing feature in Grafana and the stat panel in the last 8 years!

File: public/app/plugins/panel/stat/module.tsx (hunk 6 above)
Comment (drew08t) [reply public/app/plugins/panel/stat/module.tsx:96]: Agree!

File: packages/grafana-ui/src/components/BigValue/BigValue.tsx (hunk 4 above)
Comment (drew08t) [reply packages/grafana-ui/src/components/BigValue/BigValue.tsx:110]: I was attempting to use VizOrientation to determine an appropriate icon and font size to utilize vertical space. There is probably a better way to handle this, especially considering parentOrientation is going away. I'll look at a better approach here.

File: packages/grafana-ui/src/components/BigValue/BigValue.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (drew08t) [reply packages/grafana-ui/src/components/BigValue/BigValue.tsx:105]: Good call!

File: packages/grafana-ui/src/components/BigValue/BigValue.tsx (hunk 2 above)
Comment (drew08t) [reply packages/grafana-ui/src/components/BigValue/BigValue.tsx:119]: Yeah, agree! Good catch.

**Comment by drew08t** [comment]
> Some layout issues. esp in value only mode: image
> 
> Also unsure why BigValue cares about the VizRepeater layout being hard set to Horizontal in a few places. removing the parentOrientation in #79389
> 
> There is also a problem showing this option when we are showing row data (Value options > All values). In this mode we need to pick a field to use for the percent change value (but could be done in a follow up PR and percent change option hidden when Show is set to all Values in the mean time).
> 
> Screenshot from 2023-12-12 13-07-33
> 
> The other big problem here is color by, almost all other visualization tools that show stats with percent change color the percent change value red if negative and green if positive. Here we always use the same color as the value which looks a bit off.
> 
> But tricky, we could hard code coloring of percent change, a bit tricky to allow customization unless we make the change value be presented as a standalone field.,

I created two follow up tasks for [selecting field for percent change showing all values](https://github.com/grafana/grafana/issues/79484) (option is disabled for now) and [controlling color](https://github.com/grafana/grafana/issues/79486).

I'm hesitant to hard code color based on percent change sign as positive could represent either good or bad depending on what the data represents and likewise with negative. Perhaps we could match value color and if any thresholds or color schemes are defined, key off of the adjacent color for threshold or last number in the scheme based on trend direction. This could also be misleading and unexpected behavior however, so probably warrants further discussion and I suspect will ultimately require a set of well thought out options.

File: packages/grafana-ui/src/components/BigValue/BigValue.tsx [hunk 7]
```diff
@@ -87,23 +89,42 @@ export class BigValue extends PureComponent<Props> {
...
   render() {
-    const { onClick, className, hasLinks, theme } = this.props;
+    const { onClick, className, hasLinks, theme, textMode, parentOrientation } = this.props;
     const layout = buildLayout(this.props);
     const panelStyles = layout.getPanelStyles();
     const valueAndTitleContainerStyles = layout.getValueAndTitleContainerStyles();
     const valueStyles = layout.getValueStyles();
     const titleStyles = layout.getTitleStyles();
     const textValues = layout.textValues;
+    const percentChange = this.props.value.percentChange;
```
[the remark this answers is not in the document]
Comment (drew08t) [reply packages/grafana-ui/src/components/BigValue/BigValue.tsx:96]: Created a follow up issue [here](https://github.com/grafana/grafana/issues/79490).

**Comment by nmarrs** [comment]
> I'm hesitant to hard code color based on percent change sign as positive could represent either good or bad depending on what the data represents and likewise with negative. Perhaps we could match value color and if any thresholds or color schemes are defined, key off of the adjacent color for threshold or last number in the scheme based on trend direction. This could also be misleading and unexpected behavior however, so probably warrants further discussion and I suspect will ultimately require a set of well thought out options.

I agree with @torkelo in terms of default colors should most likely match convention. Otherwise it takes more mental effort for user to figure out whether or not percent change is positive vs negative, whereas going with conventional colors this is hard wired / effortless to understand at a glance

The "good" or "bad" context depends on the data being represented (which is external to whether or not the % change decreased or increased) - IMO we should optimize for ease of understanding % change at a glance and user can fill in the context based on data being represented. In other words - color of percent change has no deeper meaning than positive / negative (and for 0% maybe just default white / black text based on theme) 🤷‍♂️ 

For example check out the following example screenshots:
Screenshot 2023-12-13 at 5 21 25 PM
Screenshot 2023-12-13 at 5 22 27 PM
Screenshot 2023-12-13 at 5 22 36 PM

**Comment by nmarrs** [comment]
Digging into colors a bit deeper it is a bit more complex than expected - in edge case where background color clashes its not super great... though with dark background it really pops. In current form it is at least legible in all instances so I'm ok to merge as is and try and address color issue in follow-up task (🤞 before 10.3 release)

https://github.com/grafana/grafana/assets/22381771/f7245260-2aac-43e3-8bb9-a5bbf57e9272

**Comment by nmarrs** [comment]
[Took a stab at addressing changing color to be more conventional](https://github.com/grafana/grafana/pull/78250/commits/35d495136febc506cf1913c6bb65b011e6587460), ran into issues with contrast and tried applying shadow to address this but not super happy with current state 😬 other than this I think this is good to go 💯 

https://github.com/grafana/grafana/assets/22381771/819e7ce3-0d45-49c8-a0cc-8fac28f3d99c

**Review by nmarrs — APPROVED** [review]
🎸

**Comment by drew08t** [comment]
> [Took a stab at addressing changing color to be more conventional](https://github.com/grafana/grafana/pull/78250/commits/35d495136febc506cf1913c6bb65b011e6587460), ran into issues with contrast and tried applying shadow to address this but not super happy with current state 😬 other than this I think this is good to go 💯
> 
>  Screen.Recording.2023-12-13.at.7.11.54.PM.mov

Shadow definitely helps. I'll take a stab at massaging the styling for this a bit.

**Comment by drew08t** [comment]
> > [Took a stab at addressing changing color to be more conventional](https://github.com/grafana/grafana/pull/78250/commits/35d495136febc506cf1913c6bb65b011e6587460), ran into issues with contrast and tried applying shadow to address this but not super happy with current state 😬 other than this I think this is good to go 💯
> > Screen.Recording.2023-12-13.at.7.11.54.PM.mov
> 
> Shadow definitely helps. I'll take a stab at massaging the styling for this a bit.

I updated the original post to include the latest styling.

**Comment by torkelo** [comment]
Tried to explore some updates to the coloring/style logic and ended up with some refactors, instead of pushing directly this PR I opened a separate PR that targets this https://github.com/grafana/grafana/pull/79504 

I do think there is still a few things to fix in this PR, the biggest one is horizontal layout, seems a bit broken. The percent change does not trigger a name and value size recalc (which it should as it takes up space from those). Also the layout is a bit wonky with value in the middle 

 
Screenshot from 2023-12-14 12-03-55
Screenshot from 2023-12-14 12-03-50

Screenshot from 2023-12-14 12-04-01

**Review by imatwawana — APPROVED** [review]
Just added formatting to make clear that these are UI elements but otherwise, LGTM!

**Comment by nmarrs** [comment]
Current status:

We need to work on fixing scaling issue of percent change when text mode is "none"

We also need to address a better responsive design for horizontal orientation 

Color handling wise I think we are good enough for initial release :) (seems reasonable enough and most importantly legible in different edge cases)

**Comment by drew08t** [comment]
Okay, pushed a commit that applies flex wrap to the container div. This is the result on horizontal with graph. Not bad, need to move percent change down vertically though. I believe align items to center when horizontal is probably the way to go here instead of unset or flex-start.
Dec-15-2023 01-05-02

**Comment by nmarrs** [comment]
Just pushed a commit trying to improve the horizontal handling of percent change.

TLDR; having the font size / icon size have a minimum while not doing the same for title / value leads to inconsistent behavior / makes it virtually impossible to have consistent design

What are our potential options?
- Get rid of minimum size, testing this locally makes the current scaling down too tiny, but we could just set percent change to the value of the title's font size which might be reasonable
- Add minimum font size to both title / value - this will have a much higher surface area impact / modify existing stat panels which can backfire in my experience 😅
- Leave as is for time being and kick the can down the road for a future refactor - in current form it might be "good enough" to move forward

Video walkthrough
TBD

Edge case where current design breaks due to fixed min font size
Screenshot 2023-12-15 at 1 38 56 PM

**Comment by drew08t** [comment]
> Just [pushed a commit](https://github.com/grafana/grafana/pull/78250/commits/09322efe275be632784d9aa380287372abe64cbe) trying to improve the horizontal handling of percent change.
> 
> TLDR; having the font size / icon size have a minimum while not doing the same for title / value leads to inconsistent behavior / makes it virtually impossible to have consistent design
> 
> What are our potential options?
> 
> * Get rid of minimum size, testing this locally makes the current scaling down too tiny, but we could just set percent change to the value of the title's font size which might be reasonable
> * Add minimum font size to both title / value - this will have a much higher surface area impact / modify existing stat panels which can backfire in my experience 😅
> * Leave as is for time being and kick the can down the road for a future refactor - in current form it might be "good enough" to move forward
> 
> Video walkthrough
> 
>  video.mov 
> Edge case where current design breaks due to fixed min font size Screenshot 2023-12-15 at 1 38 56 PM

I think it looks good enough and we can revisit if needed.

**Comment by nmarrs** [comment]
One final note before we close out this PR:

re: adding flex wrap, this can lead to a state where only some % change wrap down while others in the same panel are not yet wrapped - not sure if there is an easy fix for this, just want to note for the future

Screenshot 2023-12-15 at 2 04 56 PM
